### ISSUE TITLE
`tuple[T, ...]` should follow subtyping rules

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5828]
Currently, pyright treats a `tuple` of indeterminate length as though it's a gradual typing type that doesn't follow subtyping rules. That is, it treats `tuple[T, ...]` as type consistent with `tuple[T, T]` and vice versa. This is different from mypy's behavior. It treats `tuple[T, ...]` as the equivalent of a union of tuples of all sizes (zero or more) that contain only type `T`. Using this definition, `tuple[T, ...]` follows normal subtyping rules. `tuple[T, T]` is a subtype of `tuple[T, ...]` but the converse is not true.

```python
def func(v1: tuple[int, int], v2: tuple[int, ...]):
    x1: tuple[int, ...] = v1  # Mypy and pyright agree this is OK
    x2: tuple[int, int] = v2  # Mypy generates an error, pyright does not
```

Pyright doesn't currently treat `tuple[T, ...]` as a gradual typing type in invariant contexts, so `list[tuple[T, ...]]` is not assignable to `list[tuple[T]]` or vice versa. This is internally inconsistent.

To address this inconsistency, I propose to adopt mypy's treatment of `tuple[T, ...]`.

This will potentially cause some churn for source bases that use pyright as a type checker. We can use mypy_primer as a guide to tell us how painful this is likely to be.

**Comment by erictraut** [comment]
The mypy_primer run confirmed my fears that this would generate new errors in some code bases. I still think it's the right thing to do though.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
